dvc/repo/experiments/remove.py

```diff
@@ -70,6 +72,26 @@ def remove(  # noqa: C901, PLR0912
         exp_ref_list.extend(exp_refs(repo.scm, git_remote))
         removed = [ref.name for ref in exp_ref_list]
 
+    if keep:
+        all_exp_refs = exp_refs(repo.scm, git_remote)  # Get all experiments
+        selected_exp_names = {ref.name for ref in exp_ref_list}  # Current selection
+        exp_ref_list = [
+            ref for ref in all_exp_refs if ref.name not in selected_exp_names
+        ]
+
+        # Handle queued experiments
```

Comment:
so, how exactly do we handle queued experiments? could you please describe the logic?

Reply:
Well, it seemed logical to me when I wrote the code, but your question makes me think I might have misunderstood something. So let's go through it to see if I got something wrong.

The idea is similar to what is done with the exp_refs. 
1. We first make a list of all queued experiments (`all_queue_entries = list(celery_queue.iter_queued())`), 
2. We get the names of those which are selected (`selected_queue_entries = {entry.name for entry in queue_entry_list}`) -- the `queue_entry_list` is built before (in the ìf exp_names:`block). 
3. We invert the selection (`queue_entry_list = [
            entry
            for entry in all_queue_entries
            if entry.name not in selected_queue_entries
        ]`)
 That allows to set the list of entries to remove to "all the entries that are in `all_queue_entries` but not in `selected_queue_entries`"
4. We create the `removed` list that will be returned at the end
5. The tasks corresponding to `queue_entry_list` (as defined in 3.) are removed from the queue at line 101 (`remove_tasks(celery_queue, queue_entry_list)`)


This still seems logical to me even after detailing all steps, so I guess either my code needs more comments (which is fine, I'd be happy to make it clearer if needed), or I missed something entirely (which is a nice opportunity to learn, I'd only need a hint to help me see what I missed).

Reply:
no, that's fine, thanks for explaining ... I'm just trying to understand the exact semantics for now

how does in general `--queue` work with `-n`, `-A`, etc ... is just a modifier: e.g. `-A --queue` removes all queued across all commits, `-n 3 --queue` removes queued _and_ attached to the last 3 commits.

or is it _or_ semantics. E.g. `-n 3 --queue` removes last 3 commits _or_ queued.

that would be helpful to clarify.

Reply:
thanks! 

>  I'd be interested to create and work on these issues if it makes sense for the project.
> To keep this PR 'atomic', I would create a new bug / new issue + PR.

yes, 100%.  if there "obvious" bugs - let's create some small followup issues and review. I'm asking here questions to better understand the `--keep` semantics and how it should handle `--queue` and other params, not for the sake of fixing the existing issues.

changing semantics - also can be an issue to discuss first

> even --keep won't be able to save any experiment in the queue if --queue is used.

so, do we need then that code that you have in this PR for handling queues - does it change anything? can it be simplified for now?

> Is --queue actually meant to clear the whole queue ?
> If so, could we make it extra clear, maybe create a new command to do so (dvc experiments clear_queue ?) ?

I think we already have it.

https://dvc.org/doc/command-reference/queue/remove#queue-remove

I would check some GH issues / PRs / ticket that introduced both to get more context. It might be redundant and we might want to deprecate it. Or we might have introduced it for convenience (so that ppl don't have to know about `dvc queue` set of commands at all - which makes sense to me to a certain degree).

> removed list will not contain all of the actually removed experiments (the queued ones would be missing)

what are the implications of this? just misleading messages at the end?

Reply:
> yes, 100%. if there "obvious" bugs - let's create some small followup issues and review.

I'll create issues for the most important points discussed here, to serve as starting points for new discussions and reviews. 

>  I'm asking here questions to better understand the `--keep` semantics and how it should handle `--queue` and other params, not for the sake of fixing the existing issues.

Sure :-) And if it brings up Ideas for nice improvements, let's just take note of these (and the rationale that goes with them) for the moment, so we don't have to go over everything again in the future :-) 
tbh, at this point, I'm more interested in the reflection process and seeing where this seemingly simple change will lead than than seeing the PR merged quickly.
 
> changing semantics - also can be an issue to discuss first
Changing semantics is never trivial. It can break things, it's always worth discussing (or at least give it some real thought). 

> > even --keep won't be able to save any experiment in the queue if --queue is used.
> 
> so, do we need then that code that you have in this PR for handling queues - does it change anything? can it be simplified for now?

Good point indeed. We can completely discard the queue thing in my code block for now. 

> > removed list will not contain all of the actually removed experiments (the queued ones would be missing)
> 
> what are the implications of this? just misleading messages at the end?
Currently, yes, only that.
I don't think there is any test to validate this, but from my quick code analysis (i.e. PyCharm > Select > Go To > Declaration or Usage), this function is only called by the CLI, so only misleading messages and only if the remove command / function is used in some weird, unclear, way.

So probably not that big of a deal right now, but, I'd fix that anyway, not because it causes a real problem, but more from a "design" and "code cleanliness" perspective, just so that the function fulfills its part of the "contract", and to avoid any issues in the future if another piece of code relied onto that result to perform something else.